### DISCUSSION TITLE
mini.ai How to setup around/inside for class constructors?

### DISCUSSION THREAD

**Discussion by yavorski** [nvim-mini/mini.nvim#792] [Q&A]
Hi,

Is it possible to configure the `around` / `inside` objects for class constructors, primarily in JavaScript/TypeScript files, similar to how it's done for functions:

```lua
f = ai.gen_spec.treesitter({ a = "@function.outer", i = "@function.inner" }, {}),
x = ai.gen_spec.treesitter({ a = "@constructor.outer", i = "@constructor.inner" }, {}),
```

I have tried to replace `function` with `constructor` but unfortunately, it didn't worked.
I would really appreciate some help with this.
Thank you!

<details>
<summary>Full MiniAi config</summary>
<div>

```lua
Lazy.use {
  "echasnovski/mini.ai",
  dependencies = {
    { "nvim-treesitter/nvim-treesitter" },
    { "nvim-treesitter/nvim-treesitter-textobjects" }
  },
  keys = {
    { "a", mode = { "x", "o" } },
    { "i", mode = { "x", "o" } },
    { "[[", desc = "Prev start @class" },
    { "]]", desc = "Next start @class" },
    { "[]", desc = "Prev end @class" },
    { "][", desc = "Next end @class" },
    { "[f", desc = "Prev start @function" },
    { "]f", desc = "Next start @function" },
    { "[F", desc = "Prev end @function" },
    { "]F", desc = "Next end @function" },
  },
  opts = function()
    local ai = require("mini.ai")
    return {
      n_lines = 420,
      custom_textobjects = {
        o = ai.gen_spec.treesitter({
          a = { "@block.outer", "@conditional.outer", "@loop.outer" },
          i = { "@block.inner", "@conditional.inner", "@loop.inner" },
        }, {}),
        c = ai.gen_spec.treesitter({ a = "@class.outer", i = "@class.inner" }, {}),
        f = ai.gen_spec.treesitter({ a = "@function.outer", i = "@function.inner" }, {}),
        x = { "constructor" },
      },
    }
  end,
  config = function(plugin, options)
    require("mini.ai").setup(options)
    require("nvim-treesitter.configs").setup({
      textobjects = {
        move = {
          enable = true,
          set_jumps = true, -- whether to set jumps in the jumplist
          goto_next_start     = { ["]]"] = "@class.outer", ["]f"] = "@function.outer" },
          goto_next_end       = { ["]["] = "@class.outer", ["]F"] = "@function.outer" },
          goto_previous_start = { ["[["] = "@class.outer", ["[f"] = "@function.outer" },
          goto_previous_end   = { ["[]"] = "@class.outer", ["[F"] = "@function.outer" },
        }
      }
    })
  end
}
```

</div>
</details>

**Comment by echasnovski** [comment]
The input for `gen_spec.treesitter()` is a table with actually used captures. The [suggested](https://github.com/echasnovski/mini.nvim/blob/8371930e8b624feabf994c3382143e83e009ed9f/doc/mini-ai.txt#L691-L693), easiest, and most currently common approach is to install [nvim-treesitter/nvim-treesitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects) and use its captures. It has a [curated list of captures it uses](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/master/CONTRIBUTING.md).

So it looks like using `'@class.outer'` and `'@class.inner'` should be the solution here. You can see how specific capture is created in the filetype's query file ([here is one for 'typescript'](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/master/queries/typescript/textobjects.scm)); it is usually more or less clear from names which part a capture defines.
